gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemBase.java

```diff
@@ -560,6 +567,77 @@ public FSDataInputStream open(Path hadoopPath, int bufferSize) throws IOExceptio
 
     return new FSDataInputStream(in);
   }
+  // Conf not used in the method for sometime, as some changes need to be made for laters
+  private ThreadPoolExecutor initThreadPools() {
+    int maxThreads = DEFAULT_MAX_THREADS.getDefault();
+    if (maxThreads < 2) {
+      logger.atWarning().log("Maximum number of threads should at least be 2.");
+      maxThreads = 2;
+    }
+    int keepAliveTime = KEEP_ALIVE_TIME.getDefault();
+    ThreadPoolExecutor unboundedThreadPool =
+        new ThreadPoolExecutor(
+            maxThreads,
+            Integer.MAX_VALUE,
+            keepAliveTime,
+            TimeUnit.SECONDS,
+            new LinkedBlockingQueue<>(),
+            BlockingThreadPoolExecutorService.newDaemonThreadFactory("gcs-transfer-unbounded"));
+    unboundedThreadPool.allowCoreThreadTimeOut(true);
+    return unboundedThreadPool;
+  }
+
+  public GoogleHadoopFileStatus getGcsFileStatus(FileStatus fileStatus) {
+    if (fileStatus instanceof GoogleHadoopFileStatus) {
+      return (GoogleHadoopFileStatus) fileStatus;
+    }
+    return null;
+  }
+
+  @Override
+  public CompletableFuture<FSDataInputStream> openFileWithOptions(
+      final Path rawPath, final OpenFileParameters parameters) throws IOException {
+    final Path path = makeQualified(rawPath);
+    Set<String> mandatoryKeys = parameters.getMandatoryKeys();
+    AbstractFSBuilderImpl.rejectUnknownMandatoryKeys(
+        mandatoryKeys, Collections.emptySet(), "for " + path);
+    FileStatus fileStatus = parameters.getStatus();
+    GoogleCloudStorageItemInfo itemInfo = null;
+    GoogleHadoopFileStatus gcsFileStatus;
+    CompletableFuture<FSDataInputStream> result = new CompletableFuture<>();
+    // Checking if fileStatus is null and calling the super implementation
+    if (fileStatus != null) {
+      gcsFileStatus = getGcsFileStatus(fileStatus);
+      if (gcsFileStatus != null) {
+        itemInfo = gcsFileStatus.getItemInfo();
+      }
+      checkNotNull(itemInfo, "Item info cannot be null");
+      logger.atFine().log("ItemInfo :%s", itemInfo);
+      logger.atFine().log("File exists: %s", itemInfo.getResourceId());
+      GoogleCloudStorageItemInfo finalItemInfo = itemInfo;
+      ThreadPoolExecutor unboundedThreadPool = initThreadPools();
+      unboundedThreadPool.submit(
+          () -> LambdaUtils.eval(result, () -> open(finalItemInfo, parameters.getBufferSize())));
+      return result;
+    } else {
+      return super.openFileWithOptions(rawPath, parameters);
+    }
+  }
+
+  protected FSDataInputStream open(GoogleCloudStorageItemInfo itemInfo, int bufferSize)
+      throws IOException {
+    checkOpen();
```

Comment:
why do we need to `checkOpen` in the `open` method ?

Reply:
The checkOpen() methods checks the existence of the GHFS or its proper initialization by checking gcsfsSupplier.